Allow percent values for speed in linear curves

DanielGibson opened the pull request:
.. in addition to the existing 0..255 system

Contributes to #387

Also allow setting `useUnscaledCurveValues: true` in fan configs, which prevents the scaling from 0..255 to MinPwm..MaxPwm. So with a PwmMap that maps 0: 0, 1: 1, ..., 255:255 the speed values from the curve are used (almost) unaltered. *("Almost" because if `neverStop` is set, they're clamped to `MinPwm, 255`)*

----

TODOs from the discussion below:

* [x] Option to disable setting fan mode every cycle
* [x] Periodic checks for whether the fan mode has been changed (and restoring it), every few seconds or so
    - [x] Option to disable those as well
    - [x] If not possible (can't read PWM) and not disabled, print a warning - once, at startup
* [x] Maybe ensure that MinPwm is 1% not 0% (unless NeverStop)?
* [x] Adapt LinearCurve steps [mapstructure stuff](https://github.com/markusressel/fan2go/pull/390#discussion_r2147140743)
* [x] LinearCurve step translation (string->float) in `LoadConfig()`?
* [x] Rework pwmMap so it really always has 256 entries and doesn't need to call `findClosestDistinctTarget()` all the time anymore

markusressel commented on internal/configuration/curves.go:
```diff
@@ -17,8 +17,9 @@ type LinearCurveConfig struct {
-	Steps map[int]float64 `json:"steps"`
+	// Steps is a map of temperature to relative speed value (in range of 0..255 or alternatively 0%..100%)
+	Steps      map[int]string
+	FloatSteps map[int]float64 `json:"steps"` // TODO: what is `json:"steps"` used for? should it be here or on Steps?
```
`json:"steps"` specifies the name of the property in JSON parsing, which is used for both the API as well as the YAML config. Although if this is already working it seems like it is not used for the Config? :thinking:

Can you change the `fan2go.yaml` to use percent values to check?
I guess this should be the "preferred" way from now on anyway.
We should definitely add tests for this though anyway.

DanielGibson replied:
That's what I assumed - but in fact the name in the yaml.config seems to depend on the literal variable name - `steps:` sets the `Steps` variable.

I was very confused by this, but the only way to get the values from the yaml to show up here was to name the new string map `Steps` (no matter where I put that `json:"steps"` thing) - originally I wanted to leave the float values in `Steps` and call the string one `StringSteps` or something, but that didn't work

markusressel replied:
Apparently [viper](https://github.com/spf13/viper) uses [mapstructure](https://github.com/go-viper/mapstructure?tab=readme-ov-file), which has its own tags:

https://pkg.go.dev/github.com/go-viper/mapstructure/v2#hdr-Renaming_Fields

DanielGibson replied:
Ok, so 
```
InSteps map[int]string `mapstructure:"user"`
Steps    map[int]float64 `json:"steps"`
```
should work?
(I guess the API should access the steps as floats?)

markusressel replied:
Maybe like this?
```golang
Steps      map[int]string      `mapstructure:"steps" json:"-"`
FloatSteps map[float64]float64 `mapstructure:"-" json:"steps"`
```

The "minus" syntax is mentioned here:
![image](https://github.com/user-attachments/assets/212d12ca-5e53-475e-84f2-de009f2bb167)

Not sure if this also works for the `json` tag, but something similar probably exists.

DanielGibson replied:
Thanks!

json also seems to use "-": https://www.digitalocean.com/community/tutorials/how-to-use-struct-tags-in-go#ignoring-private-fields

markusressel commented on internal/configuration/fans.go:
```diff
@@ -16,6 +16,13 @@ type FanConfig struct {
+	// If UseUnscaledCurveValues is set to true, the values from the curve for a specific temperature
+	// are directly mapped with PwmMap, **without** scaling them first.
+	// Note: If NeverStop is also set to true, values smaller than MinPwm are replaced with MinPwm
+	UseUnscaledCurveValues bool `json:"useUnscaledCurveValues"`
```
As commented here: https://github.com/markusressel/fan2go/issues/387#issuecomment-2945369462
I think the scaling thats currently done might be wrong in any case.

The intention was to reduce the impact of sections where changing the PWM value doesn't result in any actual RPM change. So a curve value of 
- `0` should mean "no spin"
- `1` means "spinning at `MinPwm` (minimum speed achievable on the given fan)
- `255` spinning at `MaxPwm` (maximum speed achievable on the given fan)

I am not sure the current implementation (I havent seen the changes in this PR yet) achieves this. It should for the `neverStop = true` case, but I don't think it does for the `false` case, as it will either use `0` or still use the `MinPwm` value, which would result in the fan never stopping too.

If the behavior of the scaling was changed to mean what is written above, would this remedy the need for this config option in your point of view?

Changing this would of course also be a breaking change though, so I would like to do some versioning stuff beforehand.

DanielGibson replied:
This PR changes several different things (sorry about that, but they're all related at least and many of the changes interact or depend on each other).

Most relevant for what you wrote is commit bbcd4b67749b6 ("Speeds are now always between fan.MinPwm and fan.MaxPwm"). It changes the behavior for the case that `neverStop` is **not** set - to behave almost exactly like the case where `neverStop` **is** set already does.

And this is very similar to what you wrote there:  
The speed values (between 0.0 and 255.0) are scaled to the range from `MinPwm` (minimum achievable speed) to `MaxPwm` (biggest PWM value that's supported or makes a difference in fan speed).
*(Before this change, in the non-`neverStop` case the speed values were scaled to `0` to `MaxPwm`.)*

Now the only difference between `neverStop` being set or not is that when `neverStop` is **not** set, speed value `0` (or actually values `< 1.0`) is converted to PWM 0 so the fan stops.
But all other speed values from the curve result in the same PWM value in both cases.

I don't know if this should be considered a breaking change (the fan will run a bit faster if `neverStop` is not set, especially for smaller speed values, but I think in general it won't be *radically* different, `MinPwm` should usually be pretty small anyway) - I guess you had similar changes in behavior before, when details about how MinPwm/MaxPwm work were changed (or those options were introduced) - did you change the config version then?

----

The `useUnscaledCurveValues` config option from commit e15f5e69 ("useUnscaledCurveValues option in FanConfig") is optional, if not set it defaults to false and behaves just like before this commit, meaning like I described above.

If it *is* set, it means that the speed values coming from the curve are rounded to the closest integer and then fed directly in the PWM map.
Well, *except* if `neverStop` is set, then, if the speed value from the curve is `< MinPwm`, I set it to `MinPwm` (well, actually `MinPwm + f.minPwmOffset`). But the values are not scaled.

----

Together those changes make configuring curves easier/more intuitive:
* `useUnscaledCurveValues` allows using raw PWM values in the curve, which probably some users (\*cough\* like me) already thought was happening due to the speed values being between 0 and 255 and partly even being documented as PWM values until very recently
* When **not** using that option, `neverStop` doesn't affect the fan curve anymore, except for whether 0 is treated as "stop" or not. And in this case the newly supported percent speed values make it easier to use (IMHO)

markusressel commented on internal/control_loop/direct.go:
```diff
@@ -52,10 +52,8 @@ func (l *DirectControlLoop) Cycle(target int) int {
-	rounded := int(math.Round(stepTarget))
 	// ensure we are within sane bounds
-	result := util.Coerce(rounded, 0, 255)
+	result := util.Coerce(stepTarget, 0, 255)
```
I would have thought that changing the implementation to use percentages would also change this clamping to `[0, 100]`, is this intentionally left at `255`?

DanielGibson replied:
Look more closely at the code: I didn't change the implementation to use percentages.
Internally it uses exactly the same as before, those values between 0 and 255 that are scaled to MinPwm..MaxPwm.

But if a step has the value `42%` (instead of just `42`), that is converted to a value between 0 and 255 (by multiplying it with 255/100 = 2.55).

This minimizes potential for internal breakage and ensures backwards-compatibility

markusressel commented on internal/backend.go:
```diff
@@ -283,7 +283,8 @@ func initializeFanControllers(pers persistence.Persistence, fanMap map[configura
 		}
 
-		fanController := controller.NewFanController(pers, fan, controlLoop, updateRate)
+		skipAutoPwmMapping := false // TODO: get from config? currently only used as flag of `fan2go fan init`
```
Maybe we can also name this more explicit though, f.ex. `assume-identity-pwm-map`?

DanielGibson replied:
Those things are orthogonal, as far as I can tell.

See my comment at https://github.com/markusressel/fan2go/pull/363#issuecomment-2942863909
> It prevents the automatic creation of the SetPwmToGetPwmMap in computeSetPwmToGetPwmMapAutomatically() (or rather just uses a 1:1 map there), which I think is the actually problematic part.
> However the normal PwmMap is still created based on the users configuration, if any (in computePwmMap() -> computePwmMapAutomatically())

Currently if `computeSetPwmToGetPwmMapAutomatically()` created bullshit values (due to PWM values not being immediately applied/readable), `computePwmMapAutomatically` doesn't work properly either, because it uses `f.setPwmToGetPwmMap`

DanielGibson replied:
I don't think users (should) know about the setPwmToGetPwmMap (whichs usefulness I still doubt), so I didn't want to give the option an even more confusing name.

And IIRC the proper pwmMap is calculated (even more directly than it is either way) from the setPwmToGetPwmMap if the user didn't specify any pwmMap

DanielGibson replied:
Of course suggestions for a better name for that option are welcome - I'm not completely happy with it either, but I wanted something that
* is not too long (`--skip-auto-pwm-map` is already pretty long)
* is at least somewhat understandable for the user without being too familiar with implementation details
* looks like it could help if their PWM map detection went wrong and they saw messages that suggest that the PWM map is wrong (even though they didn't specify one in the config), like  
  ```
  DEBUG   Setting PWM of cpufan to 31, found closest distinct PWM value at 31,
          applying PWM Map yields 72
  DEBUG   Setting Fan PWM of 'cpufan' to 72 ...
  DEBUG   Fan cpufan: Actual PWM value differs from requested one, skipping:
          requested: 31, expected: 72, actual: 255
  ```

markusressel replied:
Hmm ok, but in that case this also affects the "Third Party" check, if the `setPwmToGetPwmMap` is just assuming a 1:1 mapping, which might be noteworthy as well.

I am not sure it is possible to have a user not be aware of these implementation details. If we want to let a user fix a faulty "Third Party" check, the knowledge about the `setPwmToGetPwmMap` has to be gained anyway.

It looks to me as if this flag is achieving two things a the same time:
1. avoid the initial computation of the `setPwmToGetPwmMap`
2. set a 1:1 `pwmMap`

And I would want to not mix these, if possible, as they have different purposes. Yes the `pwmMap` uses the data from the `setPwmToGetPwmMap` to be computed, but thats simply because its the only guess we have (other than assuming 1:1).

It might also make sense to "warn/info log" the user that applying this flag will only persist within the db, since the override is not part of the config. I have seen people delete their db file in issues to debug previously.

My understanding that the "correct fix" for this situation would be to:
1. manually specify a 1:1 `setPwmToGetPwmMap` to get rid of the initial computation at each startup (which is currently not yet possible, but can hopefully reuse the logic proposed in #363 )
2. Optional - manually specify a 1:1 `pwmMap` in the config to (if the map derived from the now specified `setPwmToGetPwmMap` is still not correct)
3. Optional - manually disable the "third party" check in the config (if specifying the correct `setPwmToGetPwmMap` is not possible)

Of course it would be great to document all of this in detail in the README.md or at least the example config at some point.

DanielGibson replied:
> It looks to me as if this flag is achieving two things a the same time:
>
> 1. avoid the initial computation of the setPwmToGetPwmMap

yes, that's what it's for

> 2. set a 1:1 pwmMap

No - as I wrote 

> the normal PwmMap **is still created** based on the **users configuration**

I'll have to look into that 3rd-party check situation again.
It's probably a desirable feature to allow disabling it.

DanielGibson replied:
> There is (almost) no value in providing a config option that will only have effect on the very first run of fan2go, nobody will create a perfect config on first try and know that they should have set this config option, and deleting their db over and over shouldn't be the merit.

I completely disagree.
At least to me it seems very plausible that a user continues to use (and extend) a config file *without* keeping the DB around.
For example when reinstalling the system I'd keep the config, but not necessarily the .db.
Or when adding/configuring an additional fan I can imagine that people just remove the .db if they run into problems - sucks if then the old fan stops working because they forgot that last time they had to manually init it with a commandline flag.

I think the commandline flag is useful to try if this helps at all, but if it turns out to be needed I'd set the option in the config.

markusressel replied:
There might be a misunderstanding here. I don't want to prevent someone from setting a config option to achieve the desired behavior, but I don't want to introduce a setting that implicitly sets a value to something that will be exposed directly via the config as well.

The `setPwmToGetPwmMap` will be configurable via the config directly, setting a 1:1 map there will have the same effect as the `skipAutoPwmMapping` config option.

 > sucks if then the old fan stops working because they forgot that last time they had to manually init it with a commandline flag.

This can also happen if a user simply forgets and doesn't set it in the config after trying it via the CLI command. If thats an issue, not adding this flag sounds more reasonable, but I don't think thats what youre after, is it?

markusressel commented on internal/controller/controller.go:
```diff
@@ -311,53 +316,77 @@ func (f *DefaultFanController) UpdateFanSpeed() error {
+
+	// TODO: in theory even pwmTarget could be a float, because f.setPwm() looks for the closest value
+	//       in the pwm map and uses that
+	var pwmTarget int
```
Yes, I would have thought so as well. I would also remove the `pwm` wording entirely at this stage, since the `target` is not a `pwm` value anymore (semantically speaking). Only after applying the `pwmMap` does it become one again, right?

markusressel replied:
> I mean, this value is based directly on MinPwm (especially in the NeverStop case)

Yes I have also had this thought... again its all historical, because in the beginning _everything_ was a PWM value. The addition of the `file`, `cmd` and `nvidia` types "require" the internal workings of fan2go to be much more generic than it used to be, so it is a good point in time time to re-evaluate this. With the current changes, the `target` _can_ be a PWM Value in the end, but it doesn't have to be. The fan doesn't even need to use pulse width modulation anyway, so the term is kinda misleading to begin with.

F.ex. removing the `NeverStop` check from the `fan` implementations is a good step in this direction (since its non of the fan's logics business to change their reported MinPWM value based on it).

DanielGibson replied:
I think the question here is: Is PWM the value you have before applying the pwmMap, or the value you get afterwards?
Because for "real" PWM fans that map should be an identity, right?
So one could argue that the pwmMap maps a PWM value to something backend-specific

(BTW: You know what's **really** good/efficient/fast at mapping 0..255 to a value? An array/slice with 256 elements - `pwmMap map[int]int` could be `pwmMap [256]int` :slightly_smiling_face:)

markusressel replied:
I am not sure what value you are referring to when saying "PWM", in my mind there is

1. the "curve value", which has a value in 0-255, or 0-100%, or whatever. This value is based on the sensor associated with a curve, and doesn't really have a name, I usually just call it the "curve value". This is returned by `curve.Evaluate()`.
2. the "target value", which is computed by the fan controller based on the output of a curve (mentioned above), the control loop (via `controlLoop.cycle(target)`) and any additional constraints (like `NeverStop`, min/max range of the fan etc.). For "normal" fans (with a 1:1 pwmMap) this value represents a PWM value, but at this point it is still abstracted. This is triggered in `fanController.UpdateFanSpeed()`. 
3. the "PWM value", which is computed by inputting the "target" value into the `pwmMap`, and yields the actual value that is applied to the fan (I think thats what you refer to as backend-specific value", right?). This is what `fanController.setPwm(target)` does.

The fact that the curve is abstract, and that it can now (in this PR) be configured using percentage values, also would make me think that everything in step 2, like f.ex. the min/max values should get the same treatment? But then again these values are used (as you said) on a technical level to determine where the fan starts and stops, to make the curve act on a reasonable range. It would be harder to configure correctly if the fan uses 0-255 but the config in fan2go requires percentage values, although it could also support both I guess, just like in this PR. But I am still not sure if using the "non-abstracted" value in the "abstracted" part is a good way of doing it.

markusressel commented on internal/controller/controller.go:
```diff
@@ -311,53 +316,77 @@ func (f *DefaultFanController) UpdateFanSpeed() error {
+		if pwmTarget > 0 && pwmTarget < minPwm {
+			// the fan wouldn't spin with this PWM value anyway, so set 0 instead
+			// (might be better for the hardware and preserve energy)
+			pwmTarget = 0
```
:+1:

markusressel commented on internal/controller/controller.go:
```diff
@@ -311,53 +316,77 @@ func (f *DefaultFanController) UpdateFanSpeed() error {
+			// if the fan is allowed to stop and the calculated target is about 0, just set PWM 0
+			pwmTarget = 0
+		} else {
+			// other target values are mapped to minPwm..maxPwm
```
```suggestion
			// other target values are mapped to [minPwm..maxPwm]
```

markusressel commented on internal/controller/controller.go:
```diff
@@ -311,53 +316,77 @@ func (f *DefaultFanController) UpdateFanSpeed() error {
-			lastSetTargetEqualsNewTarget := lastTarget == target
-			if shouldNeverStop && lastSetTargetEqualsNewTarget {
+			lastSetTargetEqualsNewTarget := lastTarget == pwmTarget
+			if shouldNeverStop && lastSetTargetEqualsNewTarget { // TODO: also check if fan supports reading RPM?
```
This whole block is wrapped in 
```golang
if fan.Supports(fans.FeatureRpmSensor) {
    ...
}
```

So this should already be checked.

DanielGibson replied:
oh right, missed that

markusressel commented on internal/controller/controller.go:
```diff
@@ -367,8 +396,10 @@ func (f *DefaultFanController) UpdateFanSpeed() error {
 
+	// FIXME: does this really have to be called each time Pwm is set?!
+	//        couldn't it be called if SetPwm() fails?
 	_ = trySetManualPwm(f.fan)
```
> couldn't it be called if SetPwm() fails?

Sadly no, because if there is another tool interfering with fan2go it might change the control mode at any given moment, and afaik there is no way for fan2go to detect this, and there might be no way for a user to disable this third party. Of course this is an edge case, but it is an edge case that exists even in my own setup.

We might be able to check the current value first before setting it, but even that (reading the control mode) isn't 
 supported by all drivers. Its a wild wild west out there. And I am not sure if this would save many cpu cycles in the end.

Also we could add a config option for this, or only enable this if a change by a third party is detected. But this check also has its own set of gotchas, as you have also already noticed, so its not that simple either.

DanielGibson replied:
(Shocking how broken some BIOSes are)

An option (that probably is disabled by default) sounds like a good idea - I hope that this is only rarely needed...

Periodic checks sound like a good idea

markusressel commented on internal/curves/linear.go:
```diff
@@ -16,25 +17,52 @@ var (
+			} else {
+				if isPercent {
+					// convert from 1..100 percent to 0..255
+					speed = speed * 2.55
```
Ah I see, its still always working with `[0..255]` internally.
But in that case, was the change to `float` even necessary?

DanielGibson replied:
yes, because 50% is 127.5, for example

markusressel commented on internal/curves/linear.go:
```diff
@@ -16,25 +17,52 @@ var (
+			}
+			speed, err := strconv.ParseFloat(str, 64)
+			if err != nil {
+				ui.Warning("Invalid curve step value '%s' in %s", origstr, c.Config.ID)
```
This should also be validated by `fan2go config validate`

DanielGibson replied:
I don't know - can we?
The config code seems to be >90% autogenerated magic (and it's unclear to me what it's even generated from, [apparently it's not the `json:"bla"` annotations](https://github.com/markusressel/fan2go/pull/390#discussion_r2131072274)) and right now internal/configuration/curves.go only defines structs and has no code, so I'm not sure how to do this.

If you can give me a hint how to hook that into the curve config loading code, I'd be thankful, because I agree that it seems to be the best place to do this conversion (and validation)

DanielGibson commented on internal/curves/linear.go:
```diff
@@ -47,22 +75,22 @@ func (c *LinearSpeedCurve) Evaluate() (value int, err error) {
 	}
 
-	ui.Debug("Evaluating curve '%s'. Sensor '%s' temp '%.0f°'. Desired PWM: %d", c.Config.ID, sensor.GetId(), avgTemp/1000, value)
+	ui.Debug("Evaluating curve '%s'. Sensor '%s' temp '%.0f°'. Desired speed: %.2f", c.Config.ID, sensor.GetId(), avgTemp/1000, value)
```
because 50% is 127.5, 30% is 76.5 etc

DanielGibson commented:
I rebased this onto the branch for #396, assuming that it will be merged first.

I implemented the curve step validation with redundant code, but if you can tell me how to do the conversion from (string) `Steps` to `FloatSteps` in the config code I'm of course still willing to do that change.

I added a `skipAutoPwmMap` config option for fans - of course still open for a better name, just like in case of the fan init commandline option.

----

I have a question regarding using the pwmMap:  
`DefaultFanController.applyPwmMapToTarget()` does
```go
closestToTarget := f.findClosestDistinctTarget(target)
return f.pwmMap[closestToTarget]
```
Somehow findClosestDistinctTarget() feels redundant. Doesn't the pwmMap map every value between `0` and `255` to the appropriate target pwm/speed/whatever?
At least the pwmMap doc says "This map **always** contains the full range of [0..255] as keys" - what would be the point of that if not mapping all values between 0 and 255 to the most appropriate output value?

So shouldn't `return f.pwmMap[target]` be sufficient?

DanielGibson commented:
This fucking linter is driving me insane!
First it complains about using just  `%` in fmt.Errorf() - ok, fair point, that's a real bug.

Then it complains about punctuation at the end of an error string - in the same strings!
And after fixing that it complains that "error strings should not be capitalized" - whyever the fuck not, an also, couldn't it have said so last time, because it's again about the same fmt.Errorf() calls?!

markusressel commented:
> Why? If for every value between 0 and 255 the output-pwm value is calculated in the way it is now (with the **users** PWM map) and stored in a PWM map, you'd get the exact same behavior as now, but at runtime need only one lookup in the resulting PWM map instead of `findClosestDistinctTarget()`

Yes, but I am not sure what kind of interpolation corresponds to "finding the closest value". Or do you want to suggest something else?

I tried to look up issues where the `pwmMap` was relevant, and these two are probably the most interesting:

https://github.com/markusressel/fan2go/issues/161#issuecomment-1272551915

https://github.com/markusressel/fan2go/issues/311#issuecomment-2365265064

The first is a "three state" fan.
The second is a "percentage based fan", similar to the Nvidia driver.

Looking at the `pwmMap`s in these issues, I think a "step interpolation" would be what is desired, but thats not what "findClosestDistinctTarget" does, which operates more like `util.ExpandMapToFullRange()` does.

Idk what is more intuitive/logical:
This one:

```yaml
pwmMap:
  0: 0
  128: 128
  255: 255
```

where the resulting `pwmMap` is interpolated like `ExpandMapToFullRange`, "switching" between output pwm values at halfway-point between keys?

Or:

```yaml
pwmMap:
  0: 0
  64: 128
  195: 255
```

where the resulting `pwmMap` is interpolated stepwise, and the keys of the `pwmMap` provided by the user must already be the halfway-points?

Since the `pwmMap` is a very technical setting, I think I tend to say the second, which would also not require any breaking change. It would still behave differently though, as the "findClosestDistinctTarget" that would have chosen the key of `195` already at a target value of `160`, while a stepped interpolated function would only switch to `195` at this exact value (`195`). While writing this sounds more reasonable to me though, the old behavior seems misleading.

Thoughts?

DanielGibson commented:
> Yes, but I am not sure what kind of interpolation corresponds to "finding the closest value". Or do you want to suggest something else?

I'm suggesting a loop iterating from 0 to 255 and doing the exact same calculation that `findClosestDistinctTarget()` does for each of them (with the user-supplied map), and storing the result in a map that gets actually used.
The behavior would be 100% the same as currently, but without the overhead of calling `findClosestDistinctTarget()` each cycle

DanielGibson commented:
> The comparison that comes to my mind is the lighting in my home. The brightness of all of them is controlled in a percentage from [0..100]%. All of these lights turn off at 0% and on at 1%. The light output between different lights at 1% varies

I have (very) slightly tweaked the meaning of curve values once more, for the case that `UseUnscaledCurveValues` is **not** set. Now:
* `0` and `0%` mean stop (PWM 0), unless NeverStop is set, then it's set to MinPwm
* `1` and `1%` now both correspond to MinPwm (no matter if NeverStop is set or not)
* `255` and `100%` correspond to MaxPwm, like before
* => Now `1%` - `100%` gets scaled to `1..255`, instead of scaling `0%` - `100%` to `0..255`

DanielGibson commented:
I implemented an `AlwaysSetPwmMode` option for fans. If set to true, it sets the PWM Mode / ControlMode each cycle, otherwise (and by default) it doesn't.

Also added several tests for many of the new/changed features

markusressel commented:
Hey @DanielGibson thx for the update.
Sorry for the delay, lots going on right now.
I will get back to this soon, hopefully on the weekend :crossed_fingers: 

One quick note: Lets try to keep the scope of this PR as is, because it already makes it very hard for me to keep up with everything going on, especially when there are days between the sessions :sweat_smile: This in turn in turn makes it hard to review everything, merge it, and move on to the next bit. I guess this also partly contributes to the delay in response, because the time slot I need to thoroughly go through this gets longer and longer. I get that the changes are all related somehow, I just wanted to mention this to avoid misunderstandings. Thx again!

DanielGibson commented:
BTW, I think it would be good to have that periodic check (from a future PR) in version 1.0.0 - it could solve that potential "fan might not be in manual mode after suspend+resume" issue, by using real time: Even with only check once per minute, that minute would most probably be over after suspend+resume (though probably every 10 seconds or would be a saner default)

markusressel commented:
> > Yes, but I am not sure what kind of interpolation corresponds to "finding the closest value". Or do you want to suggest something else?
> 
> I'm suggesting a loop iterating from 0 to 255 and doing the exact same calculation that `findClosestDistinctTarget()` does for each of them (with the user-supplied map), and storing the result in a map that gets actually used. The behavior would be 100% the same as currently, but without the overhead of calling `findClosestDistinctTarget()` each cycle

Ok thats a good suggestion, but lets leave optimization out of the equation in this PR. I am sure there are lots of things that can be improved about performance in fan2go, lets separate it to another PR.

markusressel commented:
> > The comparison that comes to my mind is the lighting in my home. The brightness of all of them is controlled in a percentage from [0..100]%. All of these lights turn off at 0% and on at 1%. The light output between different lights at 1% varies
> 
> I have (very) slightly tweaked the meaning of curve values once more, for the case that `UseUnscaledCurveValues` is **not** set. Now:
> 
> * `0` and `0%` mean stop (PWM 0), unless NeverStop is set, then it's set to MinPwm
> * `1` and `1%` now both correspond to MinPwm (no matter if NeverStop is set or not)
> * `255` and `100%` correspond to MaxPwm, like before
> * => Now `1%` - `100%` gets scaled to `1..255`, instead of scaling `0%` - `100%` to `0..255`

Sounds good! I made a note about that for the release note draft.

markusressel commented on fan2go.yaml:
```diff
@@ -146,10 +167,11 @@ curves:
+        # Sensor value (degrees Celsius) -> Speed (0%: Fan stopped; 1%: MinPwm; 100%: MaxPwm)
+        - 40: 0%   # up to 40°C stop the fan (unless neverStop is true)
+        - 41: 1%   # at 41°C the fan should run at its minimum (non-stopped) speed
+        - 65: 33%  # at 50°C the fan should run at 33% speed
```
there is a mismatch between the value (65) and the comment, should probably be 50?

DanielGibson replied:
true

markusressel commented:
I will run the current state of the branch on two of my machines for a couple of hours or days to test things out, but first results look promising :+1:

DanielGibson commented on fan2go.yaml:
```diff
@@ -83,6 +83,27 @@ fans:
+    skipAutoPwmMap: true
+    # (Re)set the PWM mode to manual each cycle. Works around buggy BIOSes and similar
+    # that periodically overwrite fan2go's settings. Disabled by default.
+    alwaysSetPwmMode: true
```
I don't think this is currently the case.
For example
* `neverStop` defaults to false (and is optional), doesn't it?
* minPwm/startPwm/maxPwm default to 0/0/255 (or whatever fan init measured)
* pwmMap doesn't exist by default (or is something generated by fan init, or maybe a 1-to-1 mapping, depending on the situation)

"Disabled by default" implies that it's optional? But I can add "(Optional)" to these options, of course.
I just found it false to write `skipAutoPwm: false` or `alwaysSetPwmMode: false` and then describe what they do when set to true
